### PR TITLE
DNSSnooper: Support Dot1Q and limit decoding errors

### LINKED ISSUE #2155: DNSSnooper: error decoding packet: No decoder for layer type Dot1Q
```
[12:48 PM]  
carlpett So, first feedback :slightly_smiling_face: We get a lot of this:
`<probe> ERRO: 2017/01/26 12:46:27.282902 DNSSnooper: error decoding packet: No decoder for layer type Dot1Q`

[12:48 PM]  
Problem related to running on physical hardware?

[1:27 PM]  
tom googles Dot1Q...

[1:27 PM]  
tom
> Dot1Q is an IEEE's open standard, which be used to create trunk connection between switches of different vendors. In ISL the original ethernet frame is not modified, it is encapsulated between an ISL header and an FCS.
```

### PR DESCRIPTION
**Pull request by 2opremio**
Fixes #2155

I haven't been able to test this properly though, since I don't know how obtain a Dot1Q interface.

### CODE REVIEW

File: probe/endpoint/dns_snooper_linux_amd64.go [hunk 1]
```diff
@@ -203,6 +207,20 @@ func (s *DNSSnooper) run() {
 	}
 }
 
+// handleDecodeError logs errors up to the maximum allowed count
+func (s *DNSSnooper) handleDecodingError(err error) {
+	str := err.Error()
+	count := s.decodingErrorCounts[str]
+	count++
+	s.decodingErrorCounts[str] = count
```
Comment (rade) [inline probe/endpoint/dns_snooper_linux_amd64.go]: if the error string is variable, e.g. contain some id, or a timestamp, etc, etc, then this becomes a memory leak.

Comment (2opremio) [inline probe/endpoint/dns_snooper_linux_amd64.go]: I can limit the map size.

Comment (2opremio) [inline probe/endpoint/dns_snooper_linux_amd64.go]: Ideally we should use a size-limited cache with time-based expiration on keys. That's overkill though. We are running into this problem quite often, so I may consider creating a library for this.

File: probe/endpoint/dns_snooper_linux_amd64.go [hunk 2]
```diff
@@ -203,6 +208,25 @@ func (s *DNSSnooper) run() {
 	}
 }
 
+// handleDecodeError logs errors up to the maximum allowed count
+func (s *DNSSnooper) handleDecodingError(err error) {
+	// prevent potential memory leak
+	if len(s.decodingErrorCounts) > maxDecodingErrorCardinality {
```
Comment (ekimekim) [inline probe/endpoint/dns_snooper_linux_amd64.go]: optional: maybe have a flag for "reached max cardinality", so the first time we reach it we can emit a final `DNSSnooper: Too many unique error messages, silencing all errors` or similar?

Comment (2opremio) [inline probe/endpoint/dns_snooper_linux_amd64.go]: If we ever get there we would need to rethink the approach. It's a bit of a hack.

**Pull request merged**